pymatgen/entries/compatibility.py

```diff
@@ -407,9 +411,127 @@ def __str__(self):
         return "{} {} Correction".format(self.name, self.compat_type)
 
 
-class Compatibility(MSONable):
+class Compatibility(MSONable, metaclass=abc.ABCMeta):
     """
-    The Compatibility class combines a list of corrections to be applied to
+    Abstract Compatibility class, not intended for direct use.
+    Compatibility classes are used to correct the energies of an entry or a set
+    of entries. All Compatibility classes must implement .get_adjustments method.
+    """
+    @abc.abstractmethod
+    def get_adjustments(self, entry: ComputedEntry):
+        """
+        Get the energy adjustments for a ComputedEntry.
+
+        This method must generate a list of EnergyAdjustment objects
+        of the appropriate type (constant, composition-based, or temperature-based)
+        to be applied to the ComputedEntry, and must raise a CompatibilityError
+        if the entry is not compatible.
+
+        Args:
+            entry: A ComputedEntry object.
+
+        Returns:
+            [EnergyAdjustment]: A list of EnergyAdjustment to be applied to the
+                Entry.
+
+        Raises:
+            CompatibilityError if the entry is not compatible
+        """
+        return
+
+    def process_entries(self, entries: Union[ComputedEntry, list], clean: bool = False):
+        """
+        Process a sequence of entries with the chosen Compatibility scheme.
+
+        Args:
+            entries: ComputedEntry or [ComputedEntry]
+            clean: bool, whether to remove any previously-applied energy adjustments.
+                If True, all EnergyAdjustment are removed prior to processing the Entry.
+                Default is False.
+
+        Returns:
+            A list of adjusted entries.  Entries in the original list which
+            are not compatible are excluded.
+        """
+        # convert input arg to a list if not already
+        if isinstance(entries, ComputedEntry):
+            entries = [entries]
+
+        processed_entry_list = []
+
+        if clean:
+            for e in entries:
+                # remove any previously documented corrections from the entry
+                e.energy_adjustments = []
+
+        # apply the corrections
+        for entry in entries:
+            try:
+                adjustments = self.get_adjustments(entry)
+                for ea in adjustments:
+                    # Has this correction already been applied?
+                    if ea.name in [e.name for e in entry.energy_adjustments]:
+                        if ea.value in [e.value for e in entry.energy_adjustments]:
+                            # we already applied this exact correction. Do nothing.
+                            pass
+                        else:
```

Comment:
I think you're best of with a simple check that just looks for duplicate corrections
Something like:
```python
adjustment_names = [ea.name for ea in adjustments]
if len(adjustment_names) != len(set(adjustment_names)):
    raise ComptabilityError(....)
else:
    entry.energy_adjustments = adjustments
```

Reply:
The concern here is not that `get_adjustments` will return duplicate adjustment names (in fact, that's allowed b/c we assume that the implementer knows what they're doing), but rather that the adjustments returned by `get_adjustments` already exist in `ComputedEntry.parameters.energy_adjustments`. 

I want this block to ensure two behaviors when a user runs `process_entries`

1. If the exact same correction has already been applied, don't do anything. There's no need to raise a `CompatibiliityError` , as this might just have resulted from calling `process_entries` twice (e.g., user gets processed entries from MPRester and then tries to reprocess themself).

2. Raise a `CompatibilityError` if an adjustment with the same (name, cls) already exists but the values don't match. This indicates a definite problem indicative of unintended behavior or incorrect use of `Compatibility`.

Reply:
Simple code = easy to follow and test. Complex code will breed bugs and even with your checks as I mentioned above you have a bug. 

You're working too hard to "fix" issues caused by implementations. If the implementor wants a correction twice, they can use enumerated names or something. 

If you insist on checking the name and value use a `tuple`:
```python
adjustment_tuples = [(ea.name,ea.value) for ea in adjustments]
if len(adjustment_tuples) != len(set(adjustment_tuples)):
    raise ComptabilityError(....)
else:

You can put the logic for finding which ones are duplicate within the `if` which encapsulates it away. 

If you don't want process_entry to raise an error, then don't. Purposefully throwing an error to just catch it and then remove it's `Exception` quality is awkward code. Use the logger and print out an error message instead.